File: Day69/Part5/party_balancing.py
def get_cost(x, v):
    ops = 0
    while x > v + 1:
        x //= 2
        ops += 1
    if x == v or x == v + 1:
        return ops
    return -1

def solve():
    try:
        line = input().split()
        if not line: return
        n = int(line[0])
        
        lvl = list(map(int, input().split()))
        if not lvl:
            print(0)
            return
            
        min_val = min(lvl)
        
        candidates = set()
        curr = min_val
        while True:
            candidates.add(curr)
            if curr > 0:
                candidates.add(curr - 1)
            if curr == 0:
                break
            curr //= 2
        candidates.add(0)
        
        min_ops = -1
        
        for v in candidates:
            if v < 0: continue
            
            current_ops = 0
            possible = True
            
            for x in lvl:
                cost = get_cost(x, v)
                if cost == -1:
                    possible = False
                    break
                current_ops += cost
                
            if possible:
                if min_ops == -1 or current_ops < min_ops:
                    min_ops = current_ops
                    
        print(min_ops)
    except EOFError:
        pass

if __name__ == "__main__":
    solve()

File: Day69/Part5/party_balancing.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>
#include <map>

using namespace std;

// Function to get cost to reach target v or v+1 from x
// Returns -1 if impossible
int get_cost(int x, int v) {
    int ops = 0;
    while (x > v + 1) {
        x /= 2;
        ops++;
    }
    if (x == v || x == v + 1) return ops;
    return -1;
}

int main() {
    int n;
    if (cin >> n) {
        vector<int> lvl(n);
        int min_val = 2e9;
        for (int i = 0; i < n; ++i) {
            cin >> lvl[i];
            if (lvl[i] < min_val) min_val = lvl[i];
        }

        // Candidates for target base v are derived from min_val
        // min_val must eventually become some v' or v'+1
        // So v' must be reachable from min_val (as x or x-1)
        
        set<int> candidates;
        int curr = min_val;
        while (true) {
            candidates.insert(curr);
            if (curr > 0) candidates.insert(curr - 1);
            if (curr == 0) break;
            curr /= 2;
        }
        candidates.insert(0); // Always check 0

        long long min_ops = -1;

        for (int v : candidates) {
            if (v < 0) continue;
            long long current_ops = 0;
            bool possible = true;

            for (int x : lvl) {
                int cost = get_cost(x, v);
                if (cost == -1) {
                    possible = false;
                    break;
                }
                current_ops += cost;
            }

            if (possible) {
                if (min_ops == -1 || current_ops < min_ops) {
                    min_ops = current_ops;
                }
            }
        }

        cout << min_ops << endl;
    }
    return 0;
}
